app/controllers/csv_preview_controller.rb

```diff
@@ -1,13 +1,58 @@
+require "csv"
+
 class CsvPreviewController < ApplicationController
+  MAXIMUM_COLUMNS = 50
+  MAXIMUM_ROWS = 1000
+
   def show
-    legacy_url_path = request.path.sub("/preview", "").sub(/^\//, "")
     @asset = GdsApi.asset_manager.whitehall_asset(legacy_url_path).to_hash
 
+    csv_preview = CSV.parse(media, encoding:, headers: true)
+
+    @csv_rows = csv_preview.to_a.map { |row|
+      row.map { |column|
+        {
+          text: column,
+        }
+      }.take(MAXIMUM_COLUMNS)
+    }.take(MAXIMUM_ROWS + 1)
+
     parent_document_path = URI(@asset["parent_document_url"]).request_uri
     @content_item = GdsApi.content_store.content_item(parent_document_path).to_hash
 
     @attachment_metadata = @content_item.dig("details", "attachments").select do |attachment|
       attachment["url"] =~ /#{legacy_url_path}$/
     end
   end
+
+private
+
+  def legacy_url_path
+    request.path.sub("/preview", "").sub(/^\//, "")
+  end
+
+  def media
+    GdsApi.asset_manager.whitehall_media(legacy_url_path).body
+  end
+
+  def encoding
+    @encoding ||= if utf_8_encoding?
+                    "UTF-8"
+                  elsif windows_1252_encoding?
+                    "windows-1252"
+                  else
+                    raise FileEncodingError, "File encoding not recognised"
+                  end
+  end
+
+  def utf_8_encoding?
+    media.force_encoding("utf-8").valid_encoding?
+  end
+
+  def windows_1252_encoding?
+    media.force_encoding("windows-1252")
+    # This regexp checks for the presence of ASCII control characters, which
+    # would indicate we have the wrong encoding.
+    media.valid_encoding? && !media.match(/[\x00-\x08\x0b\x0c\x0e-\x1f]/)
```

Comment:
Do we need this regex check if if statement goes through utf8 first (I am assuming these ASCII control characters are utf8 but perhaps I'm wrong)?

Reply:
I've no idea TBH.  I just copied it straight from Whitehall as I wasn't sure what problem they were trying to resolve.

Rails has a lot of encoding functionality built into it now, that may not have existed when this code was first written in Whitehall.  It's tempting to remove this regex and hope Rails can deal with correctly now.

Reply:
Ugh yes I've just gone down a rabbit hole trying to trace back to the [imminence implementation](https://github.com/alphagov/imminence/pull/64/files) this was copied from and then trying to work out why that changed, but failed. Hmm. Is there any automated testing against real data we could do in integration - e.g. loop through some csvs, curl and check we get a 200 response back?

Reply:
I'm a bit in two minds here - would be nice to remove redundant code if redundant (esp since this line is a bit confusing), but also aware that if we're wrong about it being redundant this might end up with us having to re-investigate bugs we learned the answer to a long time ago 🤷‍♂️

Reply:
I'm going to remove it.  I also looked back through the commit messages and couldn't understand why the change was made.  We also don't have any examples of what problem the code was trying to fix, so I'm struggling to work out how we would be able to find an affected file for testing.

Reply:
Yeah I was thinking more take a random subset of csvs and ping them and see that they all worked without the line, but I assume that even if it was useful it'd be for a super niche case so probably wouldn't catch it anyway. Sounds good to me